src/zjs_sensor.c

```diff
@@ -264,21 +280,49 @@ static void zjs_sensor_onchange_c_callback(void *h, void *argv)
         return;
     }
 
-    union sensor_reading *reading = (union sensor_reading*)argv;
     if (zjs_sensor_get_state(handle->sensor_obj) == SENSOR_STATE_ACTIVATED) {
         zjs_sensor_update_reading(handle->sensor_obj,
                                   handle->channel,
-                                  *reading);
+                                  argv);
     }
 }
 
 // INTERRUPT SAFE FUNCTION: No JerryScript VM, allocs, or likely prints!
-static void zjs_sensor_signal_callbacks(sensor_handle_t *handle,
-                                        union sensor_reading reading)
+static void zjs_sensor_signal_callbacks(struct sensor_data *data)
 {
+    sensor_handle_t *handles = NULL;
+    bool xyz = false;
+
+    switch(data->channel) {
+    case SENSOR_CHAN_ACCEL_XYZ:
+        handles = accel_handles;
+        xyz = true;
+        break;
+    case SENSOR_CHAN_GYRO_XYZ:
+        handles = gyro_handles;
+        xyz = true;
+        break;
+    case SENSOR_CHAN_LIGHT:
+        handles = light_handles;
+        break;
+
+    default:
+        ERR_PRINT("unsupported sensor type\n");
+        return;
+    }
+
     // iterate all sensor instances to update readings and trigger event
-    for (sensor_handle_t *h = handle; h; h = h->next) {
-        zjs_signal_callback(h->id, &reading, sizeof(reading));
+    for (sensor_handle_t *h = handles; h; h = h->next) {
+        if (xyz) {
+            double xyz[3];
+            xyz[0] = data->reading.x;
+            xyz[1] = data->reading.y;
+            xyz[2] = data->reading.z;
+            zjs_signal_callback(h->id, &xyz, sizeof(xyz));
+        } else {
+            double dval = data->reading.dval;
+            zjs_signal_callback(h->id, &dval, sizeof(dval));
+        }
```

Comment:
each sensor is only a specific type, and the IPM message received from each sensor reading is different